src/client/Microsoft.Identity.Client/PlatformsCommon/Shared/DeviceAuthManager.cs

```diff
@@ -0,0 +1,146 @@
+            IDictionary<string, string> challengeData,
+            out string responseHeader)
+        {
+            string signedJwt = $"{responseJWT.GetResponseToSign()}.{Base64UrlHelpers.Encode(signedResponse)}";
+            string authToken = $"AuthToken=\"{signedJwt}\"";
+
+            responseHeader = $"PKeyAuth {authToken}, Context=\"{challengeData["Context"]}\", Version=\"{challengeData["Version"]}\"";
+        }
+
+        private X509Certificate2 FindCertificate(IDictionary<string, string> challengeData)
+        {
+            var store = new X509Store(StoreName.My, StoreLocation.CurrentUser);
+            try
+            {
+                store.Open(OpenFlags.ReadOnly);
+                var certCollection = store.Certificates;
+                if (challengeData.ContainsKey("CertAuthorities"))
+                {
+                    return FindCertificateByCertAuthorities(challengeData, certCollection);
+                }
+
+                X509Certificate2Collection signingCert = null;
+                signingCert = certCollection.Find(X509FindType.FindByThumbprint, challengeData["CertThumbprint"],
+                    false);
+                if (signingCert.Count == 0)
+                {
+                    throw new MsalException(MsalError.DeviceCertificateNotFound,
+                        string.Format(CultureInfo.CurrentCulture, MsalErrorMessage.DeviceCertificateNotFoundTemplate,
+                            "Cert thumbprint:" + challengeData["CertThumbprint"]));
+                }
+
+                return signingCert[0];
+            }
+            finally
+            {
+
+#if NETSTANDARD || WINDOWS_APP
+                store.Dispose();
+#else
+                store.Close();
+#endif
+            }
+        }
+
+        private X509Certificate2 FindCertificateByCertAuthorities(IDictionary<string, string> challengeData, X509Certificate2Collection certCollection)
+        {
+            X509Certificate2Collection signingCert = null;
+            string[] certAuthorities = challengeData["CertAuthorities"].Split(new[] { ";" },
+                StringSplitOptions.None);
+            foreach (var certAuthority in certAuthorities)
+            {
+                //reverse the tokenized string and replace "," with " + "
+                string[] dNames = certAuthority.Split(new[] { "," }, StringSplitOptions.None);
+                string distinguishedIssuerName = dNames[dNames.Length - 1];
+                for (int i = dNames.Length - 2; i >= 0; i--)
+                {
+                    distinguishedIssuerName += " + " + dNames[i].Trim();
+                }
+
+                signingCert = certCollection.Find(X509FindType.FindByIssuerDistinguishedName,
+                    distinguishedIssuerName, false);
+                if (signingCert.Count > 0)
+                {
+                    break;
+                }
+            }
+
+            if (signingCert == null || signingCert.Count == 0)
```

Comment:
this should probably be what we do above, on line 92...maybe?

Reply:
This one can be null if cert authorities collection is empty and foreach never executes.